File: go/tests/integration/manifest_malformed_test.go
package integration_test

import (
	"strings"
	"testing"

	"github.com/htxryan/c3p/tests/integration/helpers"
)

// TestManifestMalformed_* — IV/T7 translation of TS manifest-malformed.test.ts
// (PR6 #6, F2). Pinned to the Go bin's exit codes:
//   - JSON parse errors / wrong-type-field errors: exit 1 (InvalidManifest);
//     the TS bin used exit 3, but the Go classification is "user error"
//     because the manifest was authored wrong — recoverable by editing.
//   - PR16a path-traversal: exit 3 (CONFLICT-class, the spec promise).
//   - Missing-include via tilde resolving outside project: exit 3.
//   - Unknown manifest field: R36 warning, exit 0.

func TestManifestMalformed_InvalidJsonSyntax(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"bad": {Manifest: "{not: valid json", Files: map[string]string{"x.md": "x\n"}},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "bad"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("invalid json: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(strings.ToLower(r.Stderr), "invalid") {
		t.Errorf("stderr missing 'invalid': %q", r.Stderr)
	}
	if !strings.Contains(r.Stderr, "profile.json") {
		t.Errorf("stderr missing 'profile.json' path: %q", r.Stderr)
	}
}

func TestManifestMalformed_TopLevelArray(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"bad": {Manifest: "[]", Files: map[string]string{}},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "bad"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("top-level array: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(strings.ToLower(r.Stderr), "object") {
		t.Errorf("stderr missing 'object': %q", r.Stderr)
	}
}

func TestManifestMalformed_TopLevelNull(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"bad": {Manifest: "null", Files: map[string]string{}},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "bad"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("top-level null: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
}

func TestManifestMalformed_TopLevelScalar(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"bad": {Manifest: "42", Files: map[string]string{}},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "bad"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("top-level scalar: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
}

func TestManifestMalformed_NameNotString(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"bad": {Manifest: map[string]any{"name": 42}, Files: map[string]string{}},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "bad"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("name=42: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "name") {
		t.Errorf("stderr missing 'name': %q", r.Stderr)
	}
}

func TestManifestMalformed_ExtendsArray(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"bad": {Manifest: map[string]any{"extends": []string{"a"}}, Files: map[string]string{}},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "bad"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("extends=array: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "extends") {
		t.Errorf("stderr missing 'extends': %q", r.Stderr)
	}
}

func TestManifestMalformed_IncludesNonString(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"bad": {Manifest: map[string]any{"includes": []any{"valid", 42}}, Files: map[string]string{}},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "bad"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("includes non-string: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "includes") {
		t.Errorf("stderr missing 'includes': %q", r.Stderr)
	}
}

func TestManifestMalformed_TagsNonString(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"bad": {Manifest: map[string]any{"tags": []any{"good", 42}}, Files: map[string]string{}},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "bad"},
	})
	if r.ExitCode != 1 {
		t.Fatalf("tags non-string: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "tags") {
		t.Errorf("stderr missing 'tags': %q", r.Stderr)
	}
}

// PR16a: path-traversal is a CONFLICT-class hard reject (exit 3) in the
// Go bin — upgraded from the TS bin's MissingInclude behaviour.
func TestManifestMalformed_PathTraversal(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"traversal": {
				Manifest: map[string]any{"name": "traversal", "includes": []string{"../../../.ssh/config"}},
				Files:    map[string]string{"x.md": "x\n"},
			},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "traversal"},
	})
	if r.ExitCode != 3 {
		t.Fatalf("path traversal: want 3, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "../../../.ssh/config") {
		t.Errorf("stderr missing offending raw path: %q", r.Stderr)
	}
}

func TestManifestMalformed_TildeMissingInclude(t *testing.T) {
	// Tilde-form resolves to $HOME; pointing at a guaranteed-nonexistent
	// path under $HOME produces MissingInclude (exit 3) with raw "~/..."
	// preserved. We override $HOME to a fresh t.TempDir() so the test
	// doesn't accidentally pass / fail based on whatever the runner's
	// real homedir happens to contain — without this, a developer with a
	// directory named "c3p-test-nonexistent-dir-xyzzy" in $HOME would
	// silently flip the assertion.
	helpers.EnsureBuilt(t)
	fakeHome := t.TempDir()
	t.Setenv("HOME", fakeHome)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"outside": {
				Manifest: map[string]any{
					"name":     "outside",
					"includes": []string{"~/c3p-test-nonexistent-dir-xyzzy"},
				},
				Files: map[string]string{"x.md": "x\n"},
			},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "outside"},
	})
	if r.ExitCode != 3 {
		t.Fatalf("tilde missing-include: want 3, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "c3p-test-nonexistent-dir-xyzzy") {
		t.Errorf("stderr missing original include name: %q", r.Stderr)
	}
}

func TestManifestMalformed_UnknownFieldR36(t *testing.T) {
	// R36: unknown manifest field is a degraded-but-keep-going warning.
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"with_unknown": {
				Manifest: map[string]any{"name": "with_unknown", "futureField": "xyz"},
				Files:    map[string]string{"x.md": "x\n"},
			},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "use", "with_unknown"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("unknown field: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
}

File: tests/cli/integration/manifest-malformed.test.ts
/**
 * Gap closure #6 (PR6 #6, F2 epic claude-code-profiles-yhb):
 *
 * Malformed manifest variants — each one is a class of input the resolver
 * MUST reject at parse time rather than producing a degraded ResolvedPlan.
 *
 * Coverage:
 *   - invalid JSON syntax
 *   - missing required field paths (none are strictly required per R35, but
 *     wrong-type values for optional fields must be rejected)
 *   - top-level non-object (array, scalar, null)
 *   - non-string in `name`
 *   - non-string in `extends`
 *   - non-string element in `includes` array
 *   - PR16a path-traversal: `includes: ["../../../.ssh/config"]`
 *
 * Note: PR16a's hard rejection (CONFLICT-class error) is a Go-side promise
 * per docs/specs/c3p-go-migration.md §3.5. The TS bin currently classifies
 * `../...` as a `relative`-form include and tracks it as an external trust
 * notice; a true reject doesn't happen until the resolver fails to find the
 * directory (MissingInclude). This test pins the **CURRENT** TS contract:
 * a `../../../.ssh/config` include exits non-zero (because the path doesn't
 * resolve to a directory), and the error message includes the offending raw
 * input. The Go translation in IV upgrades this to a CONFLICT-class
 * PathTraversal error — see Re-decomposition trigger in F2.
 */

import { afterEach, describe, expect, it } from "vitest";

import { makeFixture, type Fixture } from "../../helpers/fixture.js";

import { ensureBuilt, runCli } from "./spawn.js";

let fx: Fixture | undefined;
afterEach(async () => {
  if (fx) await fx.cleanup();
  fx = undefined;
});

describe("gap closure #6: malformed manifest variants (PR6 #6)", () => {
  it("invalid JSON syntax → exit 3, error names the file", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        bad: { manifest: "{not: valid json", files: { "x.md": "x\n" } },
      },
    });
    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "bad"] });
    // InvalidManifestError → ResolverError → exit 3 (CONFLICT class).
    expect(r.exitCode).toBe(3);
    expect(r.stderr.toLowerCase()).toContain("invalid");
    expect(r.stderr).toContain("profile.json");
  });

  it("top-level array (not an object) → exit 3", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        bad: { manifest: "[]", files: {} },
      },
    });
    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "bad"] });
    expect(r.exitCode).toBe(3);
    expect(r.stderr.toLowerCase()).toContain("object");
  });

  it("top-level null → exit 3", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        bad: { manifest: "null", files: {} },
      },
    });
    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "bad"] });
    expect(r.exitCode).toBe(3);
  });

  it("top-level scalar (number) → exit 3", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        bad: { manifest: "42", files: {} },
      },
    });
    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "bad"] });
    expect(r.exitCode).toBe(3);
  });

  it("name is a number → exit 3, error mentions 'name'", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        bad: { manifest: { name: 42 } as unknown as Record<string, unknown>, files: {} },
      },
    });
    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "bad"] });
    expect(r.exitCode).toBe(3);
    expect(r.stderr).toContain("name");
  });

  it("extends is an array → exit 3, error mentions 'extends'", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        bad: { manifest: { extends: ["a"] } as unknown as Record<string, unknown>, files: {} },
      },
    });
    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "bad"] });
    expect(r.exitCode).toBe(3);
    expect(r.stderr).toContain("extends");
  });

  it("includes contains a non-string element → exit 3, error mentions 'includes'", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        bad: {
          manifest: { includes: ["valid", 42] } as unknown as Record<string, unknown>,
          files: {},
        },
      },
    });
    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "bad"] });
    expect(r.exitCode).toBe(3);
    expect(r.stderr).toContain("includes");
  });

  it("tags contains a non-string element → exit 3, error mentions 'tags'", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        bad: {
          manifest: { tags: ["good", 42] } as unknown as Record<string, unknown>,
          files: {},
        },
      },
    });
    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "bad"] });
    expect(r.exitCode).toBe(3);
    expect(r.stderr).toContain("tags");
  });

  // ──────────────────────────────────────────────────────────────────────
  // PR16a path-traversal — the gap-closure test #6 explicit case.
  //
  // The TS impl currently surfaces this as a MissingInclude (the resolver
  // walks `..` segments out of the project, the directory doesn't exist,
  // exits 3). Go upgrades it to a deterministic PathTraversal ConflictError
  // BEFORE filesystem touch. The exit code is 3 in both — the contract pinned
  // here is the exit code + that the offending raw string appears in stderr
  // so an operator can locate the bad manifest entry.
  // ──────────────────────────────────────────────────────────────────────
  it("PR16a: includes contains '../../../.ssh/config' → exit 3, stderr names the offending path", async () => {
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        traversal: {
          manifest: { name: "traversal", includes: ["../../../.ssh/config"] },
          files: { "x.md": "x\n" },
        },
      },
    });
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "use", "traversal"],
    });
    expect(r.exitCode).toBe(3);
    expect(r.stderr).toContain("../../../.ssh/config");
  });

  it("PR16a: includes contains tilde-form pointing outside project → resolves, then fails missing-include cleanly", async () => {
    // Tilde-form is canonically "~/..." — resolves to homedir. Pointing at a
    // path that doesn't exist must produce MissingInclude (exit 3) with the
    // raw "~/..." preserved in stderr (R37). The Go impl applies the
    // PR16a guard with a config-trusted-base allowlist; pre-1.0, the TS bin
    // resolves it and the missing-dir check rejects it.
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        outside: {
          manifest: {
            name: "outside",
            includes: ["~/c3p-test-nonexistent-dir-xyzzy"],
          },
          files: { "x.md": "x\n" },
        },
      },
    });
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "use", "outside"],
    });
    expect(r.exitCode).toBe(3);
    expect(r.stderr).toContain("c3p-test-nonexistent-dir-xyzzy");
  });

  it("unknown manifest field → R36 warning, NOT a hard fail (use still succeeds)", async () => {
    // The contrast case to all the rejects above: an unknown KNOWN-FIELD
    // shape is a degraded-but-keep-going condition per R36. Test pins that
    // contract — exit 0 with no profile content depending on the unknown
    // field, but a warning surfaced to stderr.
    await ensureBuilt();
    fx = await makeFixture({
      profiles: {
        with_unknown: {
          manifest: { name: "with_unknown", futureField: "xyz" } as unknown as Record<
            string,
            unknown
          >,
          files: { "x.md": "x\n" },
        },
      },
    });
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "use", "with_unknown"],
    });
    expect(r.exitCode).toBe(0);
  });
});
